### PR TITLE
ClassAttributesSeparationFixer - fixes & enhancements

### PR DESCRIPTION
**Pull request by SpacePossum**
closes https://github.com/FriendsOfPHP/PHP-CS-Fixer/issues/5852
closes https://github.com/FriendsOfPHP/PHP-CS-Fixer/issues/5854
closes https://github.com/FriendsOfPHP/PHP-CS-Fixer/issues/5843 (including https://github.com/FriendsOfPHP/PHP-CS-Fixer/issues/5843#issuecomment-895432985)

fixes https://github.com/FriendsOfPHP/PHP-CS-Fixer/pull/5704#issuecomment-864200146 (cc @ruudk for option `none`)

replaces https://github.com/FriendsOfPHP/PHP-CS-Fixer/pull/5869
replaces https://github.com/FriendsOfPHP/PHP-CS-Fixer/pull/5868

I haven't found BC breaks, however there is a minor change in how comments that are not related at all to a class element are handled. This is of no concern as people shouldn't just drop comments all over the place and expect these to end up somewhere.

Additionally fixes a bunch of issues with comments, include cases where the fixer had to be run multiple times before done with fixing.

Have to check how `only_if_meta` comes into play on `master` as I suspect this act near the same as `none` now.

### CODE REVIEW

**Comment by SpacePossum** [comment]
hmm seems we cannot integration/prio test using fixtures with deprecated elements, I think this should be allowed as the test might still be of value as long as the deprecated element is still supported.

**Comment by kubawerlos** [comment]
> hmm seems we cannot integration/prio test using fixtures with deprecated elements, I think this should be allowed as the test might still be of value as long as the deprecated element is still supported.

We can simply move those tests to `misc` directory, right?

**Comment by SpacePossum** [comment]
Not really... all the integration tests with fixtures don't allow deprecations, priority, misc, etc.
Adding it is also not trivial, as the deprecation might trigger when fixing from `input` to `expect`, but we always run again on the `expect` part which might not trigger a deprecation. As such we would have to allow configuration deprecation message(s) on both the `input` and `expect` parts, not sure if worth all the extra code for this single edge case. For now I take the integration test out.

**Comment by SpacePossum** [comment]
The missing deprecation option was for all fixtures, like prio, misc,etc., I've added support for it through the `settings` part

File: src/Fixer/ClassNotation/ClassAttributesSeparationFixer.php [hunk 1]
```diff
@@ -49,9 +49,6 @@ final class ClassAttributesSeparationFixer extends AbstractFixer implements Conf
      */
     public const SPACING_ONE = 'one';
 
-    private const SUPPORTED_SPACINGS = [self::SPACING_NONE, self::SPACING_ONE];
-    private const SUPPORTED_TYPES = ['const', 'method', 'property', 'trait_import'];
```
Comment (keradus) [inline src/Fixer/ClassNotation/ClassAttributesSeparationFixer.php:53]: why dropping the constants?
(like, plenty of fixers are having possible options as constants, plenty of fixers are not. maybe we should align which approach we want to take :D )

Comment (keradus) [reply src/Fixer/ClassNotation/ClassAttributesSeparationFixer.php:53]: i see slight value in having it as `@internal public const` for re-usability in tests

File: src/Fixer/ClassNotation/ClassAttributesSeparationFixer.php [hunk 2]
```diff
@@ -218,115 +203,64 @@ protected function createConfigurationDefinition(): FixerConfigurationResolverIn
                     'const' => self::SPACING_ONE,
                     'method' => self::SPACING_ONE,
                     'property' => self::SPACING_ONE,
-                    'trait_import' => self::SPACING_ONE,
+                    'trait_import' => self::SPACING_NONE,
```
Comment (keradus) [inline src/Fixer/ClassNotation/ClassAttributesSeparationFixer.php:206]: why changing the default? we had cases in past when we considered such change as bc-breaker, we were planning change of default configuration of fixer for next major always

Comment (SpacePossum) [reply src/Fixer/ClassNotation/ClassAttributesSeparationFixer.php:206]: This was added in the last patch and take as not a default and reported as `bug`, as such patching to new default 
Btw. all (OSS) repo's I checked use this CS, also it replaces the `no_extra_lines` option (to fix the conflict) which also fixes to this as default.

File: src/Fixer/ClassNotation/ClassAttributesSeparationFixer.php (hunk 1 above)
Comment (julienfalque) [reply src/Fixer/ClassNotation/ClassAttributesSeparationFixer.php:53]: I think we should avoid using these constants in tests: if the value of a constant is changed it is a BC break but the tests won't notice. Then if the constant is used only once, it can be inlined unless the meaning of its value is not obvious, in which case the constant helps make it explicit.

File: src/Fixer/ClassNotation/ClassAttributesSeparationFixer.php (hunk 2 above)
Comment (jrmajor) [reply src/Fixer/ClassNotation/ClassAttributesSeparationFixer.php:206]: This CS is also used in examples in PSR-12.

**Review by GrahamCampbell — APPROVED** [review]

File: src/Fixer/Whitespace/NoExtraBlankLinesFixer.php [hunk 3]
```diff
@@ -80,6 +81,10 @@ final class NoExtraBlankLinesFixer extends AbstractFixer implements Configurable
      */
     public function configure(array $configuration): void
     {
+        if (isset($configuration['tokens']) && \in_array('use_trait', $configuration['tokens'], true)) {
```
Comment (GrahamCampbell) [inline src/Fixer/Whitespace/NoExtraBlankLinesFixer.php:84]: do we usually do deprecations in patch releases, or reserve that for minors?

Comment (kubawerlos) [reply src/Fixer/Whitespace/NoExtraBlankLinesFixer.php:84]: good point, let's move this to 3.1

**Comment by keradus** [comment]
Thank you @SpacePossum.

**Comment by keradus** [comment]
Unfortunately I merged this PR to 3.0 branch (1st issue, I shouldn't merge deprecations into it), and then I failed to merge 3.0 -> master (2nd issue), due to non-obvious conflict resolution with changes introduced by
> feature #5704 ClassAttributesSeparationFixer - Introduce `only_if_meta` spacing option

----

I force-override 3.0 branch to restore state before merging this PRand I opened #5920 as clone PR to calmly work to address those issues.

**Comment by ruudk** [comment]
@SpacePossum You mentioned me in the PR intro related to my https://github.com/FriendsOfPHP/PHP-CS-Fixer/pull/5704#issuecomment-864200146. I'm testing 3.1.0 but I don't understand how it's addressed.

I'm using this config:
```php
'class_attributes_separation' => [
    'elements' => [
        'const'        => 'only_if_meta',
        'method'       => 'one',
        'property'     => 'only_if_meta',
        'trait_import' => 'one',
    ],
],
```

and somehow it still removes the newline between `$producer` and `$topicDecider`:
```diff
     /**
      * @var LoggerInterface&MockInterface
      */
     private MockInterface|LoggerInterface $logger;
 
     /**
      * @var MockInterface&Producer
      */
     private MockInterface|Producer $producer;
-
     private TopicDecider $topicDecider;
     private KeyDecider $keyDecider;
```

**Comment by SpacePossum** [comment]
It seems none of my fixes where merged, but hard to see after the forced changes. I try to redo these as a bug fix again (so lowest branch) and try to merge upstream (which will be very hard as we released the new `meta-only` so I cannot touch it anymore most likely)

**Comment by ruudk** [comment]
@SpacePossum I'm a bit lost in all the follow-up PR's. Is the bad merge resolved so that the meta-only feature works again on `master`?

**Comment by SpacePossum** [comment]
yes, but I think `none` acts as a better `meta-only` version, as `none` respects PHPDocs, comments _and_ type of elements, I would vote to remove `meta-only` in the next major

**Comment by ruudk** [comment]
Interesting, `none` does seem to do what I want. Although I think it's not very clear from a user perspective. That it adds newlines while I say `none`.

**Comment by SpacePossum** [comment]
I have yet to find someone who likes this :)

```php
class Foo
{
    private string $a;
    public function B{}
    # comment
    public function C
    {
    }
    use D;
    public int $d;
}
```

**Comment by julienfalque** [comment]
Dropping an option value would be a BC break. If `none` actually is a better version of `only_if_meta`, we can make `only_if_meta` an alias of `none` and deprecate `none`.
